catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/MigrationTaskManager.java

```diff
@@ -0,0 +1,173 @@
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import ddf.catalog.data.Result;
+
+/**
+ * Encapsulates all threading state for the catalog migration process and allows inspection of
+ * its progress through a public API and the Guava asynchronous call-backs.
+ * <p>
+ * Repeat calls to exportMetacardQuery({@link List<Result>} results, {@link Long} exportGroupCount)
+ * as needed to add file write tasks to the {@link java.util.concurrent.BlockingQueue} of the
+ * {@link ExecutorService}. When finished, a call to exportFinish() is required to guarantee that
+ * no errors occurred and that the executor will be shutdown properly.
+ */
+class MigrationTaskManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MigrationTaskManager.class);
+
+    private final ListeningExecutorService taskExecutor;
+
+    private final AtomicReference<Throwable> taskFailure;
+
+    private final MigrationFileWriter fileWriter;
+
+    private final CatalogMigratableConfig catalogConfig;
+
+    /**
+     * The creation of a task manager requires the passing of migratable configurations so that
+     * task behavior can be explicitly defined (i.e. where to export, how many metacards per file,
+     * and other useful data). Also initializes the {@link MigrationFileWriter} and write directory.
+     *
+     * @param config The set of configuration options to use.
+     */
+    public MigrationTaskManager(final CatalogMigratableConfig config,
+            final MigrationFileWriter fileWriter, final ExecutorService executor) {
+        this.taskExecutor = MoreExecutors.listeningDecorator(executor);
+        this.taskFailure = createAtomicReference();
+        this.catalogConfig = config;
+        this.fileWriter = fileWriter;
+    }
+
+    /**
+     * Must be called after all invocations to exportMetacardQuery. This will attempt
+     * to gracefully shutdown the executor service and handle any exceptions that might have
+     * occured in the asynchronous threads.
+     * <p>
+     * If you are using {@link MigrationTaskManager}; that is, you have constructed an instance,
+     * you must call this method to gracefully release resources and shutdown.
+     *
+     * @throws MigrationException thrown if some of the some of the metacards couldn't be exported
+     */
+    public void exportFinish() throws MigrationException {
+        LOGGER.debug("Attempting to shutdown catalog export");
+        taskExecutor.shutdown();
+        boolean isGracefulTermination = false;
+        try {
+            isGracefulTermination = taskExecutor.awaitTermination(1L, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+            LOGGER.warn("Executor was interrupted during shutdown: " + e.getMessage(), e);
+        } finally {
+            if (!isGracefulTermination) {
+                taskExecutor.shutdownNow();
+            }
+
+            checkForFailures();
```

Comment:
This however will deterministically work, since you are awaiting all tasks to be finished before checking for failures.